### ISSUE TITLE
TemporalDiscriminator and SpatialDiscriminator

### ISSUE THREAD

**Issue by clearlyzero** [hyungting/DGMR-pytorch#4]
Hello, I recently reproduced this paper, but I encountered a problem. In the discriminator, if I add the BN layer before the full connection layer, my output value will always be a fixed value. Then I use relu (1+x)+relu (1-y), and their value will always be equal to 2. This problem has puzzled me for a long time.Then I tried your discriminator and found that the same situation occurred, which led to my loss in the training process equal to 2 forever

_net = TemporalDiscriminator()
net.train()
x = torch.rand(size=(2,18,256,256))
x = net(x)
y = torch.ones(size=(2,18,256,256))
y = net(y)
print(nn.functional.relu(1 - x).mean() + nn.functional.relu(1 + y).mean())
tensor(2.0000, grad_fn=<AddBackward0>)_

**Comment by hyungting** [comment]
Hi! This is still an unsolved issue for me as well. I was wondering that is it only the temporal discriminators problem or the spatial discriminator also encounters the same problem in your experience.

**Comment by UESTC-LXD** [comment]
sorry,I replied to you late. In my experiment, both discriminators have this problem.   :(

**Comment by hyungting** [comment]
I checked the gradients of discriminators' weights this morning and found out that almost every weight's gradient is close to zero, except for the bias of the output linear layer. Do you have any idea what may cause this problem? (Note: the gradients of generator looks okay, nothing odd happens in the generator.)

**Comment by clearlyzero** [comment]
Sorry to finish replying to you. I have too many courses（天天都上网课，老师还要提问）. I think it may be backforward() in Python. You can try to change it to backforward (retain graph=True)

**Comment by clearlyzero** [comment]
> 

or delete BN layer

**Comment by UESTC-LXD** [comment]
> 
Whether gradient disappears

**Comment by UESTC-LXD** [comment]
> I checked the gradients of discriminators' weights this morning and found out that almost every weight's gradient is close to zero, except for the bias of the output linear layer. Do you have any idea what may cause this problem? (Note: the gradients of generator looks okay, nothing odd happens in the generator.)

This seems to have happened when I did the experiment last month

**Comment by UESTC-LXD** [comment]
> I checked the gradients of discriminators' weights this morning and found out that almost every weight's gradient is close to zero, except for the bias of the output linear layer. Do you have any idea what may cause this problem? (Note: the gradients of generator looks okay, nothing odd happens in the generator.)

Sorry to finish replying to you. I have too many courses（天天都上网课，老师还要提问）. I think it may be backforward() in Pytorch. You can try to change it to backforward (retain graph=True)

**Comment by hyungting** [comment]
@clearlyzero Hi! I've still training this model recently. I looked into the gradients in the discriminators and concluded that it is  mostly the output linear layers' that are dominating the output, so I added ```nn.Dropout(0.5)``` before the linear layer to deactivate several neurons while training. This did make an impact on the loss of discriminator. The losses can oscillate around value = 1 after this transformation. However, the loss value still sticks around 1 (e.g. 0.99). 
The loss curve of discriminator looks like this right now.
<img width="350" alt="image" src="https://user-images.githubusercontent.com/51833239/205848169-16fa99cf-e930-4610-841f-92cd9f14a94f.png">

**Comment by UESTC-LXD** [comment]
> 

Is the quality of the generated radar map good？

**Comment by hyungting** [comment]
The model is still not well-trained. Since the discriminator is likely to give predictions with low value (e.g. 1e-3), I changed the loss function of discriminator. The value of prediction did become larger (e.g. 0.3~0.8), however, the generator and disrcriminator still seems to be unbalanced at some point. I am now tuning the hyperparameters to make them more balanced. The prediction from the generator of this incomplete version looks like below.
<img width="957" alt="image" src="https://user-images.githubusercontent.com/51833239/206636802-563712f0-c37c-4570-848a-f877433ae57e.png">

**Comment by UESTC-LXD** [comment]
> 
![image](https://user-images.githubusercontent.com/65875405/206638286-db13c7ec-ca11-4f20-9aeb-5cefb55cd20e.png)

My training has some effect, but it is still not very good

**Comment by UESTC-LXD** [comment]
> 

![image](https://user-images.githubusercontent.com/65875405/206638562-262ba229-e9cb-4ca0-a604-dbc8c244ba87.png)
Now I have a situation that the training times have reached a certain number, and the effect is getting worse and worse

**Comment by hyungting** [comment]
Your training result looks great! The fine features seem to be well-captured.
Would you like to discuss by email? I assume you can speack mandarin. Maybe we can discuss in private in mandarin. My email is hehyunting@caece.net

**Comment by UESTC-LXD** [comment]
> Your training result looks great! The fine features seem to be well-captured. Would you like to discuss by email? I assume you can speack mandarin. Maybe we can discuss in private in mandarin. My email is [hehyunting@caece.net](mailto:hehyunting@caece.net)
ok

**Comment by bugsuse** [comment]
I have the same problem, did you solve it?

**Comment by hyungting** [comment]
@bugsuse I replaced the hinge loss function with BCE loss function to train the discriminators and finally got a much reasonable result.

**Comment by bugsuse** [comment]
@hyungting Thanks a lot! I'm going to try it!

**Comment by bugsuse** [comment]
@hyungting I have tried using `nn.BCELoss` to replace the hinge loss, but the result is not reasonable.

I changed the line:
https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L44
to
```python
self.discriminator_loss = lambda x, y: nn.BCELoss()(x, y)
```
and the lines:
https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L229-L230
to
```python
D_loss = self.discriminator_loss(fake_D_pred, true_D_pred)
```
Is this the way you replaced it? Or anything else I have overlooked?

**Comment by clearlyzero** [comment]
> @hyungting I have tried using `nn.BCELoss` to replace the hinge loss, but the result is not reasonable.
> 
> I changed the line:
> 
> https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L44
> 
> 
> to
> ```python
> self.discriminator_loss = lambda x, y: nn.BCELoss()(x, y)
> ```
> 
> and the lines:
> 
> https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L229-L230
> 
> 
> to
> ```python
> D_loss = self.discriminator_loss(fake_D_pred, true_D_pred)
> ```
> 
> Is this the way you replaced it? Or anything else I have overlooked?
When using BCE loss, you should delete the hinge loss and use BCE directly. I tested it here, but the effect is still not particularly good.

**Comment by clearlyzero** [comment]
> @hyungting I have tried using `nn.BCELoss` to replace the hinge loss, but the result is not reasonable.
> 
> I changed the line:
> 
> https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L44
> 
> 
> to
> ```python
> self.discriminator_loss = lambda x, y: nn.BCELoss()(x, y)
> ```
> 
> and the lines:
> 
> https://github.com/hyungting/DGMR-pytorch/blob/bf65bc746180f3c6ced6f3f1bdd95c72dc9e7544/DGMR/DGMRTrainer.py#L229-L230
> 
> 
> to
> ```python
> D_loss = self.discriminator_loss(fake_D_pred, true_D_pred)
> ```
> 
> Is this the way you replaced it? Or anything else I have overlooked?
When using BCE loss, you should delete the hinge loss and use BCE directly. I tested it here, but the effect is still not particularly good.

**Comment by UESTC-LXD** [comment]
When using BCE loss, you should delete the hinge loss and use BCE directly. I tested it here, but the effect is still not particularly good.


criterion = nn.BCEWithLogitsLoss() 

 s_real = criterion(S_out_real,torch.ones_like(S_out_real))
 s_fake = criterion(S_out_fake,torch.zeros_like(S_out_fake))
loss_tmp = t_real+t_fake
